### ISSUE TITLE
Geometry resize/move

### ISSUE THREAD

**Issue by nagy135** [ifreund/river#368]
On x11/bspwm I was making use of [xdo](https://github.com/baskerville/xdo) in a script that positioned window to exact size/location (width,height, x, y).

Use case: I was watching video via mpv and wanted to do other stuff as well so I had bind that put it to floating mode and positioned it at one of bottom corners as a miniature above everything (and much more).

As I understand, external program cant resize other window like this on wayland (correct me if I m wrong) so it would have to be implemented in river. I wanted to try implementing it but first I would like to ask if adding feature like this to river would be desired at all.

**Comment by novakne** [comment]
Did you look into `snap`,`move` and `resize` commands?

**Comment by nagy135** [comment]
Yes I did, what I m looking for is **exact positioning with geometry** these commands work with **delta**. Imagine you want to move currently focused window to size of 200x200 in bottom right corner. How would you script that? If I knew current size and location I can imagine writing some script I guess, not sure how to do that tho ...

**Comment by Leon-Plickat** [comment]
> i m not sure how to achieve exact size without knowing current one. Is there a way to get that information?

No, that is currently not available.

I think a `set-dimensions [width]x[height]` and a `set-position x y` command is not unreasonable (with the normal window bounds applied, of course). In the future those could potentially even be moved to the river-toplevel protocol extension.

As a workaround for now you can use the `snap` command and try to set the size via an mpv command flag.

**Comment by nagy135** [comment]
I also think there should be *instance agnostic way* to achieve this because there are programs without command flag like that.

This exact usecase could be achieved with currently possible **delta resize/move** if i knew current size so i was thinking if cli command that **dumps current state** wouldnt be more reasonable. It might be usefull in other usecases as well (bspwm could dump json with that information). Any opinions about that? It might be more general solution to my feature request.

**Comment by Leon-Plickat** [comment]
> It might be usefull in other usecases as well (bspwm could dump json with that information). Any opinions about that?

Generally features are only implemented in river if they have a good use case, so until some concrete example that can not be solved better another way is found, I am against this. Your use case can be solved far easier with commands to directly set the position and dimensions of a window on an output (note that the output dimensions are already available).

I feel like this is information clients should not have, users would only it for weird hacks. We should keep doing what other Wayland compositors do and only expose information that is strictly needed.

But if we decide to do this, it would probably be using something like the river-toplevel protocol extension I proposed (see #314). That way it is up to the client whether the information is displayed as JSON or any other format, river would just send Wayland events. If we do that, I'll probably just extend [lswt](https://git.sr.ht/~leon_plickat/lswt), which already can output data on Wayland toplevels (fancy word for windows) as JSON, to display river specific values.
